File: android/src/main/java/com/reactnative/googlefit/GoogleFitManager.java
/**
 * Copyright (c) 2017-present, Stanislav Doskalenko - doskalenko.s@gmail.com
 * All rights reserved.
 *
 * This source code is licensed under the MIT-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * Based on Asim Malik android source code, copyright (c) 2015
 *
 **/
package com.reactnative.googlefit;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.content.IntentSender;
import android.content.DialogInterface;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.util.Log;
import java.util.ArrayList;

import com.facebook.react.bridge.ActivityEventListener;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import com.google.android.gms.common.Scopes;
import com.google.android.gms.common.ErrorDialogFragment;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.common.api.Scope;
import com.google.android.gms.common.api.Status;
import com.google.android.gms.fitness.Fitness;
import com.google.android.gms.auth.api.signin.*;
import com.google.android.gms.fitness.FitnessOptions;
import com.google.android.gms.fitness.data.DataType;


public class GoogleFitManager implements
        ActivityEventListener {

    private ReactContext mReactContext;
    private static final int REQUEST_OAUTH = 1001;
    private static final String AUTH_PENDING = "auth_state_pending";
    private static final int GOOGLE_FIT_PERMISSIONS_REQUEST_CODE = 33;
    private static boolean mAuthInProgress = false;
    private Activity mActivity;

    private DistanceHistory distanceHistory;
    private StepHistory stepHistory;
    private BodyHistory bodyHistory;
    private HeartrateHistory heartrateHistory;
    private CalorieHistory calorieHistory;
    private NutritionHistory nutritionHistory;
    private StepCounter mStepCounter;
    private StepSensor stepSensor;
    private RecordingApi recordingApi;
    private ActivityHistory activityHistory;
    private HydrationHistory hydrationHistory;
    private SleepHistory sleepHistory;
    private FitnessOptions mFitnessOptions;

    private static final String TAG = "RNGoogleFit";

    public GoogleFitManager(ReactContext reactContext, Activity activity) {

        //Log.i(TAG, "Initializing GoogleFitManager" + mAuthInProgress);
        this.mReactContext = reactContext;
        this.mActivity = activity;

        mReactContext.addActivityEventListener(this);

        this.mStepCounter = new StepCounter(mReactContext, this, activity);
        this.stepHistory = new StepHistory(mReactContext, this);
        this.bodyHistory = new BodyHistory(mReactContext, this);
        this.heartrateHistory = new HeartrateHistory(mReactContext, this);
        this.distanceHistory = new DistanceHistory(mReactContext, this);
        this.calorieHistory = new CalorieHistory(mReactContext, this);
        this.nutritionHistory = new NutritionHistory(mReactContext, this);
        this.recordingApi = new RecordingApi(mReactContext, this);
        this.activityHistory = new ActivityHistory(mReactContext, this);
        this.hydrationHistory = new HydrationHistory(mReactContext, this);
        this.sleepHistory = new SleepHistory(mReactContext, this);
        //        this.stepSensor = new StepSensor(mReactContext, activity);
    }

    public GoogleApiClient getGoogleApiClient() {
        return mApiClient;
    }

    public RecordingApi getRecordingApi() {
        return recordingApi;
    }

    public StepCounter getStepCounter() {
        return mStepCounter;
    }

    public StepHistory getStepHistory() {
        return stepHistory;
    }

    public BodyHistory getBodyHistory() {
        return bodyHistory;
    }

    public HeartrateHistory getHeartrateHistory() {
        return heartrateHistory;
    }

    public DistanceHistory getDistanceHistory() {
        return distanceHistory;
    }

    public void resetAuthInProgress()
    {
        if (!isAuthorized()) {
            mAuthInProgress = false;
        }
    }

    public CalorieHistory getCalorieHistory() { return calorieHistory; }

    public NutritionHistory getNutritionHistory() { return nutritionHistory; }

    public HydrationHistory getHydrationHistory() { return hydrationHistory; }

    public SleepHistory getSleepHistory() { return sleepHistory; }

    public void authorize(ArrayList<ArrayList<String>> userScopes) {
        final ReactContext mReactContext = this.mReactContext;

        mFitnessOptions = FitnessOptions.builder()
                .addDataType(DataType.TYPE_HEART_RATE_BPM, FitnessOptions.ACCESS_READ)
                .addDataType(DataType.TYPE_WORKOUT_EXERCISE, FitnessOptions.ACCESS_READ)
                .addDataType(DataType.TYPE_WORKOUT_EXERCISE, FitnessOptions.ACCESS_WRITE)
                .build();
        GoogleSignInAccount account = GoogleSignIn.getAccountForExtension(this.mReactContext, mFitnessOptions);
        if(!isAuthorized()) {
            GoogleSignIn.requestPermissions(this.mActivity, GOOGLE_FIT_PERMISSIONS_REQUEST_CODE, account, mFitnessOptions);
        }
    }

    public void  disconnect(Context context) {
        GoogleSignInOptions options = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestEmail()
                .build();
        GoogleSignInClient googleSignInClient = GoogleSignIn.getClient(context, options);

        GoogleSignInAccount gsa = GoogleSignIn.getAccountForScopes(mReactContext, new Scope(Scopes.FITNESS_ACTIVITY_READ));
        Fitness.getConfigClient(mReactContext, gsa).disableFit();

        googleSignInClient.signOut();
    }

    public boolean isAuthorized() {
        GoogleSignInAccount account = GoogleSignIn.getAccountForExtension(this.mReactContext, mFitnessOptions);
        return  GoogleSignIn.hasPermissions(account, mFitnessOptions);
    }


    private void sendEvent(ReactContext reactContext,
                           String eventName,
                           @Nullable WritableMap params) {
        reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
    }


    @Override
    public void onActivityResult(Activity activity,int  requestCode, int resultCode, Intent data) {
        if (requestCode == GOOGLE_FIT_PERMISSIONS_REQUEST_CODE) {
            mAuthInProgress = false;
            if (resultCode == Activity.RESULT_OK) {
                // Make sure the app is not already connected or attempting to connect
                Log.d("FIT", "everything is fine");
            } else if (resultCode == Activity.RESULT_CANCELED) {
                Log.e(TAG, "Authorization - Cancel");
                WritableMap map = Arguments.createMap();
                map.putString("message", "" + "Authorization cancelled");
                sendEvent(mReactContext, "GoogleFitAuthorizeFailure", map);
            }
        }
    }

    public ActivityHistory getActivityHistory() {
        return activityHistory;
    }

    @Override
    public void onNewIntent(Intent intent) {

    }

    public static class GoogleFitCustomErrorDialig extends ErrorDialogFragment {
        @Override
        public Dialog onCreateDialog(Bundle savedInstanceState) {
            // Get the error code and retrieve the appropriate dialog
            int errorCode = this.getArguments().getInt(AUTH_PENDING);
            return GoogleApiAvailability.getInstance().getErrorDialog(
                    this.getActivity(), errorCode, REQUEST_OAUTH);
        }

        @Override
        public void onCancel(DialogInterface dialog) {
            mAuthInProgress = false;
        }
    }

    /* Creates a dialog for an error message */
    private void showErrorDialog(int errorCode) {
        // Create a fragment for the error dialog
        GoogleFitCustomErrorDialig dialogFragment = new GoogleFitCustomErrorDialig();
        // Pass the error that should be displayed
        Bundle args = new Bundle();
        args.putInt(AUTH_PENDING, errorCode);
        dialogFragment.setArguments(args);
        dialogFragment.show(mActivity.getFragmentManager(), "errordialog");
    }
}

File: android/src/main/java/com/reactnative/googlefit/GoogleFitManager.kt
package com.reactnative.googlefit
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.util.Log
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.Scopes
import com.google.android.gms.common.api.Scope
import com.google.android.gms.fitness.Fitness
import com.google.android.gms.fitness.FitnessActivities
import com.google.android.gms.fitness.FitnessOptions
import com.google.android.gms.fitness.data.DataType
import com.google.android.gms.fitness.data.DataType.TYPE_HEART_RATE_BPM
import com.google.android.gms.fitness.data.Session
import com.google.android.gms.fitness.request.SessionInsertRequest
import com.google.android.gms.tasks.OnFailureListener
import com.google.android.gms.tasks.OnSuccessListener
import java.lang.Exception
import java.util.concurrent.TimeUnit


class GoogleFitManager(private val mReactContext: ReactContext, private val mActivity: Activity) : ActivityEventListener {
    private  var mFitnessOptions = FitnessOptions.builder()
            .addDataType(TYPE_HEART_RATE_BPM, FitnessOptions.ACCESS_READ)
            .addDataType(DataType.TYPE_WORKOUT_EXERCISE, FitnessOptions.ACCESS_READ)
            .addDataType(DataType.TYPE_WORKOUT_EXERCISE, FitnessOptions.ACCESS_WRITE)
            .build()
    private lateinit var mSignInAccount: GoogleSignInAccount

    fun resetAuthInProgress() {
        if (!isAuthorized) {
            mAuthInProgress = false
        }
    }


    fun getmSignInAccount(): GoogleSignInAccount? {
        return mSignInAccount
    }
        fun saveWorkout(workout: Workout, onSuccess: OnSuccessListener<in Void>, onFailure: OnFailureListener) {
            val session: Session = Session.Builder()
                    .setName(workout.name)
                    .setDescription(workout.description)
                    .setIdentifier(workout.id)
                    .setActivity(FitnessActivities.STRENGTH_TRAINING)
                    .setStartTime(workout.startTime.toLong(), TimeUnit.MILLISECONDS)
                    .setEndTime(workout.endTime.toLong(), TimeUnit.MILLISECONDS)
                    .build()

            val insertRequest = SessionInsertRequest.Builder()
                    .setSession(session)
                    .setSession(session).build()

            Fitness.getSessionsClient(mReactContext, mSignInAccount)
                    .insertSession(insertRequest)
                    .addOnSuccessListener(onSuccess)
                    .addOnFailureListener(onFailure)
        }
    fun authorize() {
        val mReactContext = mReactContext
        mSignInAccount = GoogleSignIn.getAccountForExtension(mActivity, mFitnessOptions)
        val displayName = mSignInAccount.displayName
        if (!isAuthorized) {

            GoogleSignIn.requestPermissions(mActivity, GOOGLE_FIT_PERMISSIONS_REQUEST_CODE, mSignInAccount, mFitnessOptions)
        }

    }

    fun disconnect(context: Context?) {
        val options = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestEmail()
                .build()
        val googleSignInClient = GoogleSignIn.getClient(context!!, options)
        val gsa = GoogleSignIn.getAccountForScopes(mReactContext, Scope(Scopes.FITNESS_ACTIVITY_READ))
        Fitness.getConfigClient(mReactContext, gsa).disableFit()
        googleSignInClient.signOut()
    }

    val isAuthorized: Boolean
        get() {
            val account = GoogleSignIn.getAccountForExtension(mActivity, mFitnessOptions)
            return GoogleSignIn.hasPermissions(account, mFitnessOptions)
        }

    private fun sendEvent(reactContext: ReactContext,
                          eventName: String,
                          params: WritableMap?) {
        reactContext
                .getJSModule(RCTDeviceEventEmitter::class.java)
                .emit(eventName, params)
    }

    companion object {
        private const val GOOGLE_FIT_PERMISSIONS_REQUEST_CODE = 33
        private var mAuthInProgress = false
        private const val TAG = "RNGoogleFit"
    }

    init {
        mReactContext.addActivityEventListener(this)
    }

    override fun onNewIntent(intent: Intent?) {
    }

    override fun onActivityResult(activity: Activity?, requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == GOOGLE_FIT_PERMISSIONS_REQUEST_CODE) {
            mAuthInProgress = false

            if (resultCode == Activity.RESULT_OK) {
                // Make sure the app is not already connected or attempting to connect
                Log.d("FIT", "everything is fine")
            } else if (resultCode == Activity.RESULT_CANCELED) {
                val requestedScopes =  mSignInAccount.requestedScopes
                val grantedScopes = mSignInAccount.grantedScopes
                Log.e(TAG, "Authorization - Cancel")
                val map = Arguments.createMap()
                map.putString("message", "" + "Authorization cancelled")
                sendEvent(mReactContext, "GoogleFitAuthorizeFailure", map)
            }
        }
    }
}